pretyping/inductiveops.mli

```diff
@@ -211,10 +211,6 @@ val make_case_or_project :
 val make_default_case_info : env -> case_style -> inductive -> case_info
 i*)
 
-val compute_projections : Environ.env -> inductive -> (constr * types) array
```

Comment:
I am not sure this is a good idea to remove this function from the API. It was somewhat nice to have for plugins, and it is bound to be reimplemented (with its share of bugs) next time we need to compute projections for something else in the codebase.

Reply:
This function is actually already contained in `Record.declare_projections`. Would you be ok that it is `record.ml` which exports a `compute_projections` function?

Alternatively, the code in `Record.declare_projections` could be split into two parts, a part which computes projections in `inductiveops.ml` and a part which declares the projections, remaining in `record.ml` (or moved to `declareInd.ml` actually), for both primitive and non-primitive projections?

Reply:
I think I prefer the second option. We should not put generic code for term manipulation in vernac.

Reply:
I agree. Then, there remains the question in `inductiveops.ml` of factorizing or not the code for primitive and non-primitive projections. There is a slight difference: the type of projection `i` refers to the _predictible_ internal name `Proj(... mind ... i ...)` in the primitive case while in the non-primitive case, it refers to a name `Const(cst)` where `cst` is only (in principle) known lately from `declare_constant`. (We can cheat though, and assume that `cst` will precisely follow the rule of being of the `Projection.Repr.constant` form, the same way the code that currently uses `Projection.Repr.constant` already cheats by anticipating that the non-compact defined projection associated to a primitive projection and obtained by `declare_constant` will have this name.)

PS: I would not like to diverge too much, but afaicj the `Proj` nodes have still no concrete syntax associated to them, so, we are actually defining types of projections which refer to terms parsable only up to delta-conversion.

Reply:
@ppedrot, I see two possibilities:
- if we want to keep the type of `Inductiveops.compute_projections` exactly as it is and making it work for non-primitive projections. However, I'm not too enthousiastic about preserving exactly `Inductiveops.compute_projections` for the reasons given above that it is morally interleaved with the declaration of the `Const` names in the environment.
- An alternative, which concentrates on the technical and error-prone computation of the body of the projection of index `proj_arg` is to have a function `compute_projection_body : env -> subst:constr option list -> inductive -> ~proj_arg:int -> (constr * types) option`  where `None` in the two occurrences of `option` is to tell which previous projections (in `subst`) were not definable and to tell that the demanded projection cannot be defined, respectively (I use here labels for explanation purpose, not to be kept in code).

In both cases, `FakeRecord` would now carry a default name for inhabitants of the record, as it is done for `PrimRecord`.

Does it fit your view? Or else, what was your own idea?

Reply:
Please don't add information to `FakeRecord`, I have been trying to get rid of it for a long time and was blocked due to interaction with the semantics of functors. This stuff shouldn't be in the kernel at all, this is really due to libobject quirks. Currently IIRC it's only relied on by extraction.

I don't think the Inductiveops function should work over non-primitive records. It has been designed precisely for this use case, and is a low-level utility function that is hard to get right, just as you wouldn't reimplement say DeBruijn substitution everywhere. Why can't you split depending on the primitive status in the upper layers?

Reply:
The original reason I came to this PR is that I'm investigation how to attach manual implicit arguments to `declare_constant` rather than having `declare_constant` first declaring the automatic implicit arguments, then `declare_manual_implicit` overriding this computation with a new implicit object including the manual implicit arguments. So, my motivations were: more locality, less side-effects, more intuitive.

Now, along this exploration, I stumbled on `declare_constant` in `Record.declare_projections` which is not done at the same time in the primitive and in the non-primitive case. If found this a difficult invariant to follow, so I tried to move the `declare_constant` from `declareInd.ml` to `Record.declare_projections`. In the way, I realized that the code for non-primitive projections in `declare_projections` was the same as the code of primitive projections in `inductiveops.ml`. So, I factorized it and the resulting code was indeed simpler.

So, to answer precisely this question:

> Why can't you split depending on the primitive status in the upper layers?

There is already a split depending on the primitive status, but the fact is that the split is not on the technical part involving de Bruijn indices. The split is only on whether we use a `Proj` node or a `Case` node in the body and a `Proj` node or `Const` node in the dependencies.

----

Now, I guess that the underlying question is where we are going. If you consider `FakeRecord` to be temporary, I guess that this means that you consider the code in `Record.declare_projections` to be eventually removed and all declarations introduced by the keyword `Record` based on primitive projections, right? Then, what about `Recordops.register_primitive_projection`? I guess that the `Const` bound to `Proj` nodes are also deemed to disappear, right? And that we shall have a proper user syntax for the `Proj` node instead, right? And a proper `Proj` node in `GlobRef.t` so that we can e.g. attach a scope to its argument? And that we shall eta-expand unapplied projections at parsing time?

In this case, if the point is to keep `Inductiveops.compute_projections` for future use, that's ok to me.

Reply:
I agree with your plan, except maybe for the η-expansion of projections at parsing time. I'd lean instead towards prohibiting the use of non-fully applied projections. In languages like ML this is even enforced syntactically, and in Coq we're paying a huge cost of trying to strike a middle ground on this topic.

Reply:
Do we renounce to any form of compatibility with existing code using "positive" records, such as taking benefit of primitive projections for `prod` or `sigT`?

Reply:
I was under the impression that the conclusion of the current mess is that compatibility is but a dream...

Reply:
The original motivation for primitive projections was efficiency. In practice, it introduced two features. On one side, it brought negative record types (i.e. sugar for `match` and eta-conversion). On the other side, it brought compact representation of projections (i.e. dropping the parameters in the internal representation). I think that the two concepts should be kept separated.

That `match t with (x,y) => term end` in primitive positive records compatibility mode was sugar for `let x := fst t in let y := snd t in term` and this has been a source of problems. This introduced confusion in printing and broke reduction. So, let's renounce to simulating positive record types from negative record types and let's instead preserve `match` in positive records as it is now.

Let us then consider different levels.

The mininum servicing is just to adopt a compact representation for the constants associated to `match`-based projections in positive records. For that we need a constructor say `CompactProj(cst,t)`, thought as `App (Const cst, inferred_parameters @[t])`, and we instrument `mkApp` so that it uses `CompactProj` when applied to such constant. In all other respects, this behaves as current positive records. The bet is that all what has to be done for such a `CompactProj` is already mostly done in the code for the `Proj` node, so that `CompactProj(cst,t)` could instead be encoded, say, as `Proj(Projection.make(...,true),t)` (where we recycle `true` in the proj node to mean  projection from a "positive" type rather than "unfolded"). I believe we would mostly have compatibility up to the exception that the record parameters are not any more part of the internal syntax and thus things like `pattern t at n` would work differently. More generally, printing would be a bit different: the parameters are inferred and the user has no more control on them.

[I suspect that this approach may eventually generalize to other compact representations of application nodes in cases where some arguments are canonically inferable.]

The next level is to support eta-conversion in positive record types. The claim (to be checked formally) is that conversion modulo match-eta-conversion in positive record types is equivalent to conversion modulo proj-eta-conversion in negative record types. Thus in the conversion, we stop treating a positive record type as positive but treat it as if it were negative. This means blocking the reduction of `Proj` and expanding `match t with (x,y) => u end` into `u[x:=fst t][v:=snd t]`, as well as applying the same eta-conversion for (finite) as we already do for negative record types. At this level, I don't think we would lose a lot of compatibility, except in some cases where a more powerful conversion may matter.

A further level would be to avoid losing the compact representation of a positive projection in reduction tactics by blocking the expansion of projections in positive record types to a `match`, though still allowing its iota-contraction when applied to a constructor. This may be strange in the sense that both `fst z` and `match z with (x,y) => x end` would be normal forms and none of `cbv`, `cbn`, etc would move one to the other. We could add however a `cbv+` which prefers expanding `fst` into `match` and a `cbv-` which prefers expanding `match` into `fst` and `snd`. At this level, compatibility is virtually lost whenever we try to `unfold` a positive projection applied to a neutral term, or more generally, whenever we ask for a delta reduction of projection which cannot be followed by a iota.

If we are interesting in this last level, we are actually not far from a convergence between (finite) negative or positive record types. For instance, we could go as far as saying that `match` is also available in negative records. This would not be a big step since something behaving like `match` already exists in negative record type under the form of `let x:=fst z in let y:=snd z in t`. The only difference is that we would provide a primitive representation for it and block its reduction when `z` is not a pair. At the reduction level, `cbv`, `cbn` and co would be different but `cbv-` would implement reduction as if `match` were indeed pure sugar. No change in the way conversion is implemented though. 

In this last level, we would not need any more the `bool` flag from above to differentiate between positive and negative types. They would both behave the same in conversion and the only difference between a (finite) negative record type and a (finite) positive record type would be at the level of the reduction: `cbv+` would implement reduction as if in a positive type; `cbv-` would implement reduction as if in a negative type.